src/components/ha-service-control.ts

```diff
@@ -48,28 +50,60 @@ export class HaServiceControl extends LitElement {
 
   @property({ reflect: true, type: Boolean }) public narrow!: boolean;
 
+  @property({ type: Boolean }) public showAdvanced?: boolean;
+
   @internalProperty() private _serviceData?: ExtHassService;
 
+  @internalProperty() private _checkedKeys = new Set();
+
   @query("ha-yaml-editor") private _yamlEditor?: HaYamlEditor;
 
   protected updated(changedProperties: PropertyValues) {
     if (!changedProperties.has("value")) {
       return;
     }
+    const oldValue = changedProperties.get("value") as
+      | undefined
+      | this["value"];
+
+    if (oldValue?.service !== this.value?.service) {
+      this._checkedKeys = new Set();
+    }
+
     this._serviceData = this.value?.service
       ? this._getServiceInfo(this.value.service)
       : undefined;
 
     if (
       this._serviceData &&
       "target" in this._serviceData &&
-      this.value?.data?.entity_id
+      (this.value?.data?.entity_id ||
+        this.value?.data?.area_id ||
+        this.value?.data?.device_id)
     ) {
+      const target = {
+        ...this.value.target,
+      };
+
+      if (this.value.data.entity_id && !this.value.target?.entity_id) {
+        target.entity_id = this.value.data.entity_id;
+      }
+      if (this.value.data.area_id && !this.value.target?.area_id) {
+        target.area_id = this.value.data.area_id;
+      }
+      if (this.value.data.device_id && !this.value.target?.device_id) {
+        target.device_id = this.value.data.device_id;
+      }
+
       this.value = {
         ...this.value,
-        target: { ...this.value.target, entity_id: this.value.data.entity_id },
+        target,
+        data: { ...this.value.data },
       };
+
       delete this.value.data!.entity_id;
+      delete this.value.data!.device_id;
+      delete this.value.data!.area_id;
     }
 
     if (this.value?.data) {
```

Comment:
At this point we know `this.value` has content, the `?` is not needed

Reply:
No, `value` could be undefined here? All the checks above are in a different if statement.

Reply:
You set it with  `data`on line 98

Reply:
But that is in the if block from line 77?

Reply:
Ah, right!